tools/swift-plugin-server/Sources/swift-plugin-server/WasmEngine.swift

```diff
@@ -0,0 +1,91 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift open source project
+//
+// Copyright (c) 2024 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See http://swift.org/LICENSE.txt for license information
+// See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import WASI
+import WasmTypes
+import SystemPackage
+
+typealias WasmFunction = ([UInt32]) throws -> [UInt32]
+
+protocol WasmEngine {
+  init(path: FilePath, imports: WASIBridgeToHost) throws
+
+  func function(named name: String) throws -> WasmFunction?
+}
+
+typealias DefaultWasmPlugin = WasmEnginePlugin<DefaultWasmEngine>
+
+// a WasmPlugin implementation that delegates to a WasmEngine
+struct WasmEnginePlugin<Engine: WasmEngine>: WasmPlugin {
+  private let hostToPlugin: FileDescriptor
+  private let pluginToHost: FileDescriptor
+  private let pumpFunction: WasmFunction
+  let engine: Engine
+
+  init(path: FilePath) throws {
+    let hostToPluginPipes = try FileDescriptor.pipe()
+    let pluginToHostPipes = try FileDescriptor.pipe()
+    self.hostToPlugin = hostToPluginPipes.writeEnd
+    self.pluginToHost = pluginToHostPipes.readEnd
+
+    let bridge = try WASIBridgeToHost(
+      stdin: hostToPluginPipes.readEnd,
+      stdout: pluginToHostPipes.writeEnd,
+      stderr: .standardError
+    )
+    engine = try Engine(path: path, imports: bridge)
+
+    let exportName = "swift_wasm_macro_v1_pump"
+    guard let pump = try engine.function(named: exportName) else {
+      throw WasmEngineError(message: "Wasm plugin has an unknown ABI (could not find '\(exportName)')")
+    }
+    self.pumpFunction = pump
+
+    guard let start = try engine.function(named: "_start") else {
+      throw WasmEngineError(message: "Wasm plugin does not have a '_start' entrypoint")
+    }
+    _ = try start([])
+  }
+
+  func handleMessage(_ json: [UInt8]) throws -> [UInt8] {
+    try withUnsafeBytes(of: UInt64(json.count).littleEndian) {
+      _ = try hostToPlugin.writeAll($0)
+    }
+    try hostToPlugin.writeAll(json)
+
+    _ = try pumpFunction([])
```

Comment:
I don't understand why we're passing in and returning unused `[UInt64]` instances to the pump function. Why pass anything if we aren't going to use it anywhere?

Reply:
This is vestigial from when we were passing and returning JSON via function arguments instead of stdio; will remove.